torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+import logging
+from typing import Tuple, Any, List, Dict
+
+import torch
+from torch.fx.node import map_aggregate
+
+from .quantization_state import (
+    AutoQuantizationState,
+)
+from .utils import (
+    trace_with_inputs,
+    is_leaf,
+    pack_weights_for_functionals,
+    HookType,
+    get_torch_function_hook_type,
+    get_module_hook_type,
+)
+from . import auto_trace_rewriter
+
+logger = logging.getLogger('auto_trace')
+logging.basicConfig(level=logging.DEBUG)
+# logging.basicConfig(level=logging.INFO)
+
+# enabling this tanks performance, make sure to disable for benchmarking
+# TODO(future PR): clean this up
+enable_logging = False
+# enable_logging = True
+
+
+def add_auto_observation(
+    model : torch.nn.Module,
+    example_inputs: Tuple[Any],
+    input_dtypes: Any = (torch.float,),  # must be same structure as model inputs
+    output_dtypes: Any = (torch.float,),  # must be same structure as model outputs
+) -> torch.nn.Module:
+    def convert_to_interception_proxy(x):
+        if isinstance(x, torch.Tensor):
+            return x.as_subclass(QuantizationInterceptionProxy)  # type: ignore[arg-type]
+        else:
+            return x
+
+    cur_module = None
+    first_call = True
+    module_stack : List[torch.nn.Module] = []
+    # Counter for tensor IDs, will be modified inplace by quant state.
+    # This is used to track tensors from output ops to input ops. For example,
+    # if op_n had a tensor output with id=1, and op_n+2 had a tensor input with
+    # id=1, we know that the output of op_n is the input to op_n+2.
+    qtensor_id = [0]
+    module_id_to_fqn: Dict[int, str] = {}
+
+    class QuantizationInterceptionProxy(torch.Tensor):
+        """
+        An override of `torch.Tensor` to enable dynamic tracing for
+        quantization.
+
+        For each function with a `__torch_function__` override, this proxy does
+        the following for functions which need quantization:
+
+        1. calls `_auto_quant_state.validate_cur_op` to validate that
+           the currently seen op is the same as what was recorded during tracing
+        2. calls `_auto_quant_state.op_prepare_before_hook`
+        3. executes the original function
+        4. calls `_auto_quant_state.op_prepare_after_hook`
+        5. calls `_auto_quant_state.mark_cur_op_complete` to increment
+           the current op index in preparation for the next op
+
+        Otherwise, calls the original function.
+        """
+
+        @classmethod
+        def __torch_function__(cls, func, types, args=(), kwargs=None):
+            # to prevent printing things from going into an infinite loop
+            if func == torch.Tensor.__repr__:
```

Comment:
What is the purpose of this if statement here?

Reply:
this is for debugging, if the if statement is removed then printing tensors goes into an infinite loop.  I haven't debugged this thoroughly, would be good to do it in a follow-up.